### ISSUE TITLE
alternative ways to have upstream loaders declare ICSS deps and exports

### ISSUE THREAD

**Issue by jquense** [webpack-contrib/css-loader#1217]
### Feature Proposal

`css-loader` already supports the `:import` and `:export` syntax for declaring CSS dependencies. This is a somewhat limiting API for a few reasons:

- only allows exporting simple values, e.g. serializable ones. For preprocessors with more complex types, it'd be nice to export lists, maps, etc
- Requires the user configure ICSS support or otherwise indicate the format via `.icss.css` extension.

### Feature Use Case

I'm working on a CSS preprocesor that is Sass-like but natively understands CSS modules. At the moment i have it output ICSS in the compiled CSS, which works great when handing it off to css-loader. However the above limitations make it less nice. I also have to do weird things, like `:import(./file) { ___nothing: a }` for imports that are only side effects, otherwise css-loader removes the import as unused.

My proposal is we have `css-loader` check `meta` for ICSS data, it can use instead of checking the source string. This should be backwards compatible but make it easier for other things to tell css-loader about the exports-to-js and the dependencies of a CSS file so it can bundle them together, instead of the preprocessor.

something like: 
```js
// my loader
function (src) {
  const cb = this.async()

  const {css, exports, dependencies } = myPreprocessr.render(src)

  cb(css, map, { icss: { dependencies, exports }} )
}
```

What do you think?

**Comment by jquense** [comment]
to clarify for the dependencies, the goal would be to tell css-loader that it should include a require for that file, i don't think it makes sense to specify replacements to the css in this API, the upstream loader can just do that themselves. What it can't do is get the import to show up in JS

**Comment by evilebottnawi** [comment]
Sorry out of scope `css-loader`, you can implement this on own side

**Comment by jquense** [comment]
i can't implement this on my own side that's why i'm proposing it...

**Comment by evilebottnawi** [comment]
meta is not safe for caching and we should avoid it, you can implement own loader, you tried to go far beyond scope of `css-loader`

**Comment by jquense** [comment]
i'm sorry what am i being blamed for? I've not done anything to css-loader except talk to you about figuring out how we might be able to _remove_ functionality from it by moving css-modules out. Something you were very supportive of i thought

**Comment by evilebottnawi** [comment]
> i'm sorry what am i being blamed for?

No blame.

> I've not done anything to css-loader except talk to you about figuring out how we might be able to remove functionality from it by moving css-modules out. Something you were very supportive of i thought

Future CSS modules will be moved out of css-loader inside plugin (maybe around new year), so I don't want to add anything new here, especial meta and other abilities, the only thing we can using to solve problems is the syntax, we can improve it

**Comment by jquense** [comment]
> Future CSS modules will be moved out of css-loader inside plugin (maybe around new year)

I would love to see this. Is there any place to see early versions or designs to provide feedback/input? I am doing a lot fo exploration and work in the css-modules space (as well as others), and could provide helpful real world limits, issues and rough edges with the current css-modules approach and API's. If only to help reduce the scope and enable other authors to build things on top of it

> the only thing we can using to solve problems is the syntax, we can improve it

I think we've probably hit the limits of what we can do there in css-loader, at least not without breaking lots of people. I think exploring syntax improvements is good (i'm building a new preprocessor for css-modules!). I'm happy to wait on this if larger reworks are happening.

**Comment by jquense** [comment]
(the webpack docs should **really** mention that `meta` is unsafe to rely on)

**Comment by evilebottnawi** [comment]
> I would love to see this. Is there any place to see early versions or designs to provide feedback/input? I am doing a lot fo exploration and work in the css-modules space (as well as others), and could provide helpful real world limits, issues and rough edges with the current css-modules approach and API's. If only to help reduce the scope and enable other authors to build things on top of it

Right now no, it is on roadmap only

The logic will be pretty much the same, the only thing that will be in great possibilities is to register your own plugins, so if you want something new you will need plugin

**Comment by evilebottnawi** [comment]
> (the webpack docs should really mention that meta is unsafe to rely on)

It is unsafe in not all cases, but this case is edge case

**Comment by jquense** [comment]
@evilebottnawi slightly related. Would you be ok with allowing an alternatice css syntax for specifying ICSS exports and imports? The pseudo selector version breaks in a lot of contexts, whereas 

```
@icss-import {

**Comment by evilebottnawi** [comment]
Somewhere I have already seen this request, you we can do it, but we should support both syntax

**Comment by evilebottnawi** [comment]
Should be easy https://github.com/css-modules/icss-utils/blob/master/src/extractICSS.js

**Comment by alexander-akait** [comment]
@jquense Can we close?
